src/Repository/ProfileField.php

```diff
@@ -0,0 +1,301 @@
+				$this->aclFormatter->toString($profileFieldInputs['new']['group_allow'] ?? ''),
+				$this->aclFormatter->toString($profileFieldInputs['new']['contact_deny'] ?? ''),
+				$this->aclFormatter->toString($profileFieldInputs['new']['group_deny'] ?? '')
+			);
+
+			$newProfileField = $this->insert([
+				'uid' => $uid,
+				'label' => $profileFieldInputs['new']['label'],
+				'value' => $profileFieldInputs['new']['value'],
+				'psid' => $psid,
+				'order' => $profileFieldOrder['new'],
+			]);
+
+			$profileFieldInputs[$newProfileField->id] = $profileFieldInputs['new'];
+			$profileFieldOrder[$newProfileField->id] = $profileFieldOrder['new'];
+
+			$profileFields[] = $newProfileField;
+		}
+
+		unset($profileFieldInputs['new']);
+		unset($profileFieldOrder['new']);
+
+		// Prunes profile field whose label has been emptied
+		$profileFields = $profileFields->filter(function (Model\ProfileField $profileField) use (&$profileFieldInputs, &$profileFieldOrder) {
+			$keepModel = !isset($profileFieldInputs[$profileField->id]) || !empty($profileFieldInputs[$profileField->id]['label']);
+
+			if (!$keepModel) {
+				unset($profileFieldInputs[$profileField->id]);
+				unset($profileFieldOrder[$profileField->id]);
+				$this->delete($profileField);
+			}
+
+			return $keepModel;
+		});
+
+		// Regenerates the order values if items were deleted
+		$profileFieldOrder = array_flip(array_keys($profileFieldOrder));
+
+		// Update existing profile fields from form values
+		$profileFields = $profileFields->map(function (Model\ProfileField $profileField) use ($uid, &$profileFieldInputs, &$profileFieldOrder) {
+			if (isset($profileFieldInputs[$profileField->id]) && isset($profileFieldOrder[$profileField->id])) {
+				$psid = $this->permissionSet->getIdFromACL(
+					$uid,
+					$this->aclFormatter->toString($profileFieldInputs[$profileField->id]['contact_allow'] ?? ''),
+					$this->aclFormatter->toString($profileFieldInputs[$profileField->id]['group_allow'] ?? ''),
+					$this->aclFormatter->toString($profileFieldInputs[$profileField->id]['contact_deny'] ?? ''),
+					$this->aclFormatter->toString($profileFieldInputs[$profileField->id]['group_deny'] ?? '')
+				);
+
+				$profileField->psid  = $psid;
+				$profileField->label = $profileFieldInputs[$profileField->id]['label'];
+				$profileField->value = $profileFieldInputs[$profileField->id]['value'];
+				$profileField->order = $profileFieldOrder[$profileField->id];
+
+				unset($profileFieldInputs[$profileField->id]);
+				unset($profileFieldOrder[$profileField->id]);
+			}
+
+			return $profileField;
+		});
+
+		return $profileFields;
+	}
+
+	/**
+	 * Migrates a legacy profile to the new slimmer profile with extra custom fields.
+	 * Multi profiles are converted to ACl-protected custom fields and deleted.
+	 *
+	 * @param array $profile Profile table row
+	 * @throws \Exception
+	 */
+	public function migrateFromProfile(array $profile)
```

Comment:
```suggestion
	public function migrateFromLegacyProfile(array $profile)
```
Just to make it clear - don't use it normally ;-)

Reply:
Fair enough, done.